pkg/controllers/provisioning/controller.go

```diff
@@ -31,124 +33,125 @@ import (
+		}
+		if p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+			for _, term := range p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+				errs = multierr.Append(errs, validateNodeSelectorTerm(term))
+			}
+		}
 	}
-	// Update the provisioner if anything has changed
-	if c.hasChanged(ctx, provisioner) {
-		c.Delete(provisioner.Name)
-		c.provisioners.Store(provisioner.Name, NewProvisioner(ctx, provisioner, c.kubeClient, c.coreV1Client, c.cloudProvider))
+	return errs
+}
+
+func validatePodAffinityTerm(term v1.PodAffinityTerm) error {
+	if !v1alpha5.ValidTopologyKeys.Has(term.TopologyKey) {
+		return fmt.Errorf("unsupported topology key in pod affinity, %s not in %s", term.TopologyKey, v1alpha5.ValidTopologyKeys)
 	}
 	return nil
 }
 
-// Returns true if the new candidate provisioner is different than the provisioner in memory.
-func (c *Controller) hasChanged(ctx context.Context, provisionerNew *v1alpha5.Provisioner) bool {
-	oldProvisioner, ok := c.provisioners.Load(provisionerNew.Name)
-	if !ok {
-		return true
-	}
-	hashKeyOld, err := hashstructure.Hash(oldProvisioner.(*Provisioner).Spec, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
-	if err != nil {
-		logging.FromContext(ctx).Fatalf("Unable to hash old provisioner spec: %s", err)
+func validateNodeSelectorTerm(term v1.NodeSelectorTerm) (errs error) {
+	if term.MatchFields != nil {
+		errs = multierr.Append(errs, fmt.Errorf("node selector term with matchFields is not supported"))
 	}
-	hashKeyNew, err := hashstructure.Hash(provisionerNew.Spec, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
-	if err != nil {
-		logging.FromContext(ctx).Fatalf("Unable to hash new provisioner spec: %s", err)
+	if term.MatchExpressions != nil {
+		for _, requirement := range term.MatchExpressions {
+			if !v1alpha5.SupportedNodeSelectorOps.Has(string(requirement.Operator)) {
+				errs = multierr.Append(errs, fmt.Errorf("node selector term has unsupported operator, %s", requirement.Operator))
+			}
+		}
 	}
-	return hashKeyOld != hashKeyNew
-}
-
-// List active provisioners in order of priority
-func (c *Controller) List(ctx context.Context) []*Provisioner {
-	provisioners := []*Provisioner{}
-	c.provisioners.Range(func(key, value interface{}) bool {
-		provisioners = append(provisioners, value.(*Provisioner))
-		return true
-	})
-	sort.Slice(provisioners, func(i, j int) bool { return provisioners[i].Name < provisioners[j].Name })
-	return provisioners
+	return errs
 }
 
-// Register the controller to the manager
 func (c *Controller) Register(_ context.Context, m manager.Manager) error {
 	return controllerruntime.
 		NewControllerManagedBy(m).
 		Named(controllerName).
-		For(&v1alpha5.Provisioner{}).
-		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		For(&v1.Pod{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10_000}).
```

Comment:
Why is Add() still a blocking call?  We don't use the result, and if it wasn't this could go down to a more reasonable number.

Reply:
Let's catch this on a followup that cleans up the batcher.